6_predict/src/plot_diagnostics.R

```diff
@@ -261,6 +261,66 @@ plot_pred_obs <- function(df_pred_obs, metric, region, out_dir,
 }
 
 
+make_residual_map <- function(df_pred_obs, sites_g2_sf, metric, pred_gage_ids, region, out_dir,
+                              from_predict = FALSE, model_wf = NULL, pred_data = NULL){
+  #' @description this function creates maps of residuals (obs - predicted) and a
+  #' variogram plot
+  #' 
+  #' @param df_pred_obs is a data frame with observed (obs) and predicted (.pred) from 
+  #' the output of  predict_test_data function. if left null and from_predict is true, 
+  #' it will predict it from the model_wf (copied these lines from the plot_pred_obs function)
+  #' @param sites_g2_sf  gagesii spatial ata object with LAT LON and ID
+  #' @param metric  metric that is being predicted
+  #' @param pred_gage_ids vector of gage ids that correspond to each row of the df_pred_obs
+  #' @param region  region of the model/prediction
+  #' @param directory where output figures are saved
+  #' 
+
+  if(from_predict){
+    #predict from provided workflow and data
+    df_pred_obs <- predict(model_wf, new_data = pred_data, type = 'numeric') %>%
+      mutate(obs = pred_data[[metric]])
+  }
+  
+  lat_lons<- sites_g2_sf %>%
+    filter(ID %in% pred_gage_ids) %>%
+    select(ID, LAT, LON)
+  
+  df<- bind_cols(lat_lons,df_pred_obs) %>%
+    mutate(resid = obs - .pred)
+  
+  states <- map_data("state")
+  
+  p1<-ggplot(states, aes(x=long, y=lat, group=group)) +
+    geom_polygon(fill="white", colour="gray") +
+    geom_sf(data = df, inherit.aes = FALSE, 
+            aes(color = .data[["resid"]]), 
+            size = 0.75)+
+    scale_color_scico(palette = 'batlow')+
```

Comment:
We should use a diverging colorbar for the residuals (scico has those) and force the center to be 0. We may want to also force that the min and max values for the colorbar are the same.

For residuals, I like to use a 2-color colorbar with different shades of each color

Reply:
@jds485  I've been playing around with color scales and one problem is that the distribution of the residuals is very long tailed, so when I map it and center it at zero, basically the only thing you see are the extreme outliers and it doesn't seem super helpful:
![image](https://user-images.githubusercontent.com/18313247/181823427-74bbdad6-af10-4aab-95ea-89e311180060.png)

One thing I could do is set the gradient limit to some fixed values and then anything out of those bounds would just get the maximum value color.  Below I set the limits at the 10th and 90th percentiles.  I haven't quite figured out how to change the legend labels so that it shows <=6 and >=6 or something, but I supposed it can be done.  
![image](https://user-images.githubusercontent.com/18313247/181824134-5cbd435e-5311-4adb-87de-fc84abaf43f9.png)


Anyway - let me know what you think or if you have a different way to map them.

Reply:
I like the second option better, and that's the method I expected we would need to use. If we use white as 0, then we should make the basemap of the states a dark color (I think the scico package is visually optimal when using black as the background, so could try using black).

Reply:
ok here is an example with the updated maps and variogram.  I tried putting both the variogram and robust variogram on there but it was causing me more time and headaches than I can deal with on a Friday, the scaling of the three plots on 1 figure was just haywire and I tried many solutions and they all failed.  So - if it is important to have both variograms, I can print them out together in a separate figure from the map, otherwise this is what the plots look like.  I think I got all the comments addressed in the latest commit.  If this figure looks ok, I will merge.

![image](https://user-images.githubusercontent.com/18313247/181837807-4de7d7de-5a6f-45f7-93a8-52cc5e0b5e30.png)

Reply:
Thanks! Did you check if the units are in degrees or km? It looks like it could be degrees based on having larger residuals in the West located about 30 degrees from lower values in the East. 

These plots look good to me.

Reply:
The units are still in lat long but I don't think that is a problem for the map because the lat long units shouldn't affect the residuals.  I'm not exactly sure what the different scenarios are, so I'm not sure why the west does so much poorer here.  

The variogram I think might still be messed up because of the units though.  There is an argument in the variogram function, you can set project=FALSE but when I do this I get an error. So I am thinking maybe I need to project the lat long data prior to the variogram function but I am not sure how to do this.  The examples I've found online are all rather esoteric and I don't really know what I'm doing.

Reply:
ok - I think I have the variogram right now.  I added the  crs to the pq_sites_g2_sf target and this is how the variogram looks now:  I'm pretty sure these are in km now.  

![image](https://user-images.githubusercontent.com/18313247/182228583-65f38f4b-9fe7-4a8a-8e49-b1969a5e044c.png)

Reply:
Thanks, that makes sense to me looking at the values we have. I'm not worried about the < 20 km variability here because we have neighboring residuals with large differences in value. A semivariance of 25 corresponds to an average difference of about 7 and that's believable to me.

100 km is about 1 degree. Because our map is 10 degree tick marks, using 1000 km as the max range could be better visually.

Reply:
ok - here it is with a 1000 km max:

![image](https://user-images.githubusercontent.com/18313247/182232101-d78a0587-0a8a-4ce8-a610-848e1c201988.png)